client/daemon/rpcserver/subscriber.go

```diff
@@ -182,51 +201,54 @@ loop:
 		case info := <-s.PieceInfoChannel:
 			s.Infof("receive piece info, num: %d, finished: %v", info.Num, info.Finished)
 			// not desired piece
-			if s.totalPieces > -1 && uint32(info.Num) < nextPieceNum {
+			if uint32(info.Num) < nextPieceNum {
 				continue
 			}
 
 			s.Lock()
-			total, err := s.sendExistPieces(uint32(info.Num))
+			err := s.sendExistPieces(uint32(info.Num))
 			if err != nil {
 				err = s.saveError(err)
 				s.Unlock()
 				return err
 			}
-			if total > -1 && s.totalPieces == -1 {
-				s.totalPieces = total
-			}
-			if s.totalPieces > -1 && len(s.sentMap)+int(s.skipPieceCount) == int(s.totalPieces) {
-				s.Unlock()
-				break loop
-			}
-			if info.Finished {
+
+			nextPieceNum = s.searchNextPieceNum(nextPieceNum)
+			s.Debugf("update desired next piece num: %d", nextPieceNum)
+
+			if info.Finished && s.isAllPiecesSent(nextPieceNum) {
 				s.Unlock()
 				break loop
 			}
-			nextPieceNum = s.searchNextPieceNum(nextPieceNum)
 			s.Unlock()
 		case <-s.Success:
 			s.Infof("peer task is success, send remaining pieces")
 			s.Lock()
```

Comment:
I'm not sure if this lock is necessary? which varible does this lock protect?  totalPieces?